### PR TITLE
RBAC Domain services project update cancel

### PR DESCRIPTION
**Pull request by lancewf**
cancel
https://drive.google.com/file/d/1r1ErO5eh6EBwiWo0dHdA3yqUHqU8o3Nq/view?usp=sharing

### :nut_and_bolt: Description

Adding the ability to cancel running project updates on the domain services. 

The below diagram shows the process flow of canceling a running project update. After the update process is started, a request to cancel comes in from the API to the authz-service. The authz-service publishes a project update cancel event. Both the ingest-service and the compliance-service are listening for project update cancel events. When these services receive a project update cancel events they send a request to Elasticsearch to stop currently running tasks. After the Elasticsearch tasks are canceled, job complete status events will be published. As before the authz-service will get these status events and mark the update process complete. 

image

### :athletic_shoe: Demo Script / Repro Steps
1. `build components/ingest-service && build components/compliance-service && build components/event-service && build components/authz-service`
1. `start_all_services`
1. `chef_load_nodes 10000`
1. Watch the logs `sl &`
1. Create a project `chef-automate dev grpcurl authz-service -- chef.automate.domain.authz.v2.Projects.CreateProject -d '{"name": "project9", "id": "project9"}'`
1. Update a project `chef-automate dev grpcurl authz-service -- chef.automate.domain.authz.v2.Projects.UpdateProject -d '{"name": "project9", "id": "project9"}'`
1. Cancel the update `chef-automate dev grpcurl authz-service -- chef.automate.domain.authz.v2.Projects.ProjectUpdateCancel`
1. Confirm that the update was canceled. Watch the logs and the statuses from the authz-service, compliance-service, and ingest-service with these commands. To watch compliance update process`chef-automate dev grpcurl compliance-service -- chef.automate.domain.compliance.ingest.ingest.ComplianceIngester.ProjectUpdateStatus`, 
To watch ingest update process, `chef-automate dev grpcurl ingest-service -- chef.automate.domain.ingest.EventHandler.ProjectUpdateStatus`
To watch authz update process, `chef-automate dev grpcurl authz-service -- chef.automate.domain.authz.v2.Projects.ProjectUpdateStatus`

### :chains: Related Resources

https://github.com/chef/automate/issues/60

### :white_check_mark: Checklist

- [x] Necessary tests added/updated?
- [ ] Necessary docs added/updated?
- [x] Code actually executed?
- [x] Vetting performed (unit tests, lint, etc.)?

### CODE REVIEW

File: components/ingest-service/projectupdater/manager.go [hunk 1]
```diff
@@ -52,6 +51,26 @@ func NewManager(client backend.Client, authzProjectsClient iam_v2.ProjectsClient
...
 
+func (manager *Manager) Cancel(projectUpdateID string) {
+	switch manager.state {
+	case notRunningState:
+		// do nothing job is not running
+	case runningState:
+		if manager.projectUpdateID == projectUpdateID {
+			logrus.Debugf("Cancelling project tag update for ID %q elasticsearch task ID %q",
+				manager.projectUpdateID, manager.esJobID)
+			manager.client.JobCancel(context.Background(), manager.esJobID)
```
Comment (lancewf) [inline components/ingest-service/projectupdater/manager.go]: Tell elasticsearch to cancel the update process. Let the `waitingForJobToComplete` process publish a complete status when the job is stopped.

File: components/event-service/server/server.go [hunk 2]
```diff
@@ -63,6 +64,7 @@ func (s *Server) initRegistry() map[string][]string {
 	registry[NodeTerminated] = []string{config.CFG_KEY}
 	registry[ProjectRulesUpdate] = []string{config.CFG_KEY, config.COMPLIANCE_INGEST_KEY}
 	registry[ProjectRulesUpdateStatus] = []string{config.AUTHZ}
+	registry[ProjectRulesCancelUpdate] = []string{config.CFG_KEY, config.COMPLIANCE_INGEST_KEY}
```
Comment (lancewf) [inline components/event-service/server/server.go]: Registering ingest-service and compliance-service for project rules cancel update events.

File: components/authz-service/server/v2/project_update_manager.go [hunk 3]
```diff
@@ -269,6 +288,30 @@ func (manager *ProjectUpdateManager) startProjectUpdateForDomainResources(
 	return err
 }
 
+func (manager *ProjectUpdateManager) cancelProjectUpdateForDomainResources() error {
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: The only thing needed to cancel a project update for the domain services is to publish a 'cancel' event.

File: components/authz-service/server/v2/project_update_manager.go [hunk 4]
```diff
@@ -319,18 +362,6 @@ func (manager *ProjectUpdateManager) resetDomainServicesData() {
...
-	for _, domainService := range manager.domainServices {
-		if !domainService.complete {
-			complete = false
-			break
-		}
-	}
-
-	return complete
-}
-
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager.go]: This was not being used anymore 🤷‍♂️

File: components/authz-service/server/v2/project_update_manager_test.go [hunk 5]
```diff
@@ -10,48 +9,61 @@ import (
...
+	"github.com/golang/mock/gomock"
 	"github.com/golang/protobuf/ptypes"
 	_struct "github.com/golang/protobuf/ptypes/struct"
 	"github.com/stretchr/testify/assert"
-	"google.golang.org/grpc"
 )
 
 func TestProjectUpdateManagerOneUpdateRunningAtATime(t *testing.T) {
-	mockEventServiceClient := &MockEventServiceClient{}
+	numberOfPublishedEvents := 0
```
Comment (lancewf) [inline components/authz-service/server/v2/project_update_manager_test.go]: Updating to us the a2 service mock objects.

File: components/authz-service/server/v2/project.go [hunk 6]
```diff
@@ -166,6 +166,15 @@ func (s *state) ProjectUpdateStatus(ctx context.Context,
 	}, nil
 }
 
+func (s *state) ProjectUpdateCancel(ctx context.Context,
+	req *api.ProjectUpdateStatusReq) (*api.ProjectUpdateCancelResp, error) {
+	err := s.projectUpdateManager.Cancel()
+	if err != nil {
+		log.Errorf("Could not Cancel Project Update")
```
Comment (phiggins) [inline components/authz-service/server/v2/project.go]: Should this log message include `err`?

Comment (lancewf) [reply components/authz-service/server/v2/project.go]: Yes, that would be good to add. Thanks.

File: components/compliance-service/ingest/ingestic/ingestic.go [hunk 7]
```diff
@@ -463,6 +463,17 @@ func (backend *ESClient) UpdateSummaryProjectsTags(ctx context.Context, projectT
...
+func (backend *ESClient) JobCancel(ctx context.Context, jobID string) error {
+	_, err := elastic.NewTasksCancelService(backend.client).
+		TaskId(jobID).
+		Do(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
```
Comment (phiggins) [inline components/compliance-service/ingest/ingestic/ingestic.go]: The last 5 lines of this function could be replaced with `return err`.

File: components/compliance-service/ingest/projectupdater/manager.go [hunk 8]
```diff
@@ -53,6 +53,33 @@ func NewManager(client *ingestic.ESClient, authzProjectsClient iam_v2.ProjectsCl
...
+		// do nothing job is not running
+	case runningState:
+		if manager.projectUpdateID == projectUpdateID {
+			logrus.Debugf("Cancelling project tag update for ID %q elasticsearch task ID %v",
+				manager.projectUpdateID, manager.esJobIDs)
+
+			for _, esJobID := range manager.esJobIDs {
+				err := manager.client.JobCancel(context.Background(), esJobID)
+				if err != nil {
+					logrus.Errorf("Failed to canceled Elasticsearch task. "+
```
Comment (phiggins) [inline components/compliance-service/ingest/projectupdater/manager.go]: s/canceled/cancel/

File: components/ingest-service/backend/elastic/elastic.go [hunk 9]
```diff
@@ -64,6 +64,17 @@ func (es *Backend) JobStatus(ctx context.Context, jobID string) (backend.JobStat
...
 
+func (es *Backend) JobCancel(ctx context.Context, jobID string) error {
+	_, err := elastic.NewTasksCancelService(es.client).
+		TaskId(jobID).
+		Do(ctx)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
Comment (phiggins) [inline components/ingest-service/backend/elastic/elastic.go]: This could also be `return err`.

**Pull request merged**